### PR TITLE
Feature/strategy pattern analyzer

### PR DESCRIPTION
**Pull request by dylandreimerink**
### Veranderingen in dit pull request:
- De strategy pattern analyzer gemaakt
- ClassPart class is nu hernoemd naar FilePart omdat dat en accuratere naam is
- Happy path en edge case unit tests toegevoegd tot een redelijke code coverage

### CODE REVIEW

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+     * @param deceleration the deceleration to be set
+     * @return this
+     */
+    public ClassOrInterface setDeceleration(final ClassOrInterfaceDeclaration deceleration) {
+        this.deceleration = deceleration;
+        return this;
+    }
+
+    @Override
+    public String toString() {
```
Comment (fabianvdbor) [inline src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: Is er een reden waarom je de toString definieert en wordt deze ergens gebruikt?

File: src/main/java/nl/windesheim/codeparser/FilePart.java [hunk 2]
```diff
@@ -47,8 +49,16 @@ public Range getRange() {
...
      * @return instance of this
      */
-    public ClassPart setRange(final Range range) {
+    public FilePart setRange(final Range range) {
         this.range = range;
         return this;
     }
+
+    @Override
+    public String toString() {
```
Comment (fabianvdbor) [inline src/main/java/nl/windesheim/codeparser/FilePart.java]: Is er een reden waarom je de toString definieert en wordt deze ergens gebruikt?

File: src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Resolves FileParts.
+ */
+public final class FilePartResolver {
+
+    /**
+     *
+     */
+    private FilePartResolver() {
```
Comment (fabianvdbor) [inline src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java]: Kan je de lege constructor verwijderen of er documentatie boven plaatsen?

File: src/main/java/nl/windesheim/codeparser/patterns/Strategy.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     * The file part which contains the context class of the strategy pattern.
+     */
+    private ClassOrInterface context;
+
+    /**
+     * The file part which contains the strategy interface of the strategy pattern.
+     */
+    private ClassOrInterface strategyInterface;
+
+
```
Comment (fabianvdbor) [inline src/main/java/nl/windesheim/codeparser/patterns/Strategy.java]: Kan je deze regel hier weghalen zodat er maar één lege regel staat?

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java (hunk 1 above)
Comment (dylandreimerink) [reply src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: Ja, want de debugger geeft een preview van de toString als je een variabel inspecteert. Dus die had ik voor m'n eigen gemak toegevoegd

**Comment by dylandreimerink** [comment]
Ik denk dat het het beste is dat we jouw branch als eerste naar develop toe mergen. Dan kunnen we alles wat gefixt moet worden in develop oplossen. Of ik kan eerst develop in deze branch mergen en het hier fixen. Het ligt er maar net aan wat beter uitkomt

File: src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java (hunk 3 above)
Comment (dylandreimerink) [reply src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java]: Nee, als je die weg haalt gaat codestyle over de zijk. Je moet een constructor hebben en die moet een javadoc hebben. :(

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/EligibleStrategyContextFinder.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                try {
+                    resolvedType = JavaParserFacade.get(typeSolver).convertToUsage(variableType);
+                } catch (UnsolvedSymbolException e) {
+                    System.err.println("Can't resolve symbol: " + variableType.asString()
+                            + ", can be caused by missing dependencies, "
+                            + "invalid java code or selection of invalid source root");
+                    continue;
+                }
+
+                //if the type is a reference, not a primitive type
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/EligibleStrategyContextFinder.java]: Verplaatsen naar een aparte methode. Dan kun je ook echt unit testen door alleen de ResolvedReferenceType te testen

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/EligibleStrategyContextFinder.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                if (!(resolvedType instanceof ResolvedReferenceType)) {
+                    continue;
+                }
+
+                ResolvedReferenceTypeDeclaration typeDecleration
+                        = ((ResolvedReferenceType) resolvedType).getTypeDeclaration();
+
+                //If the type is a interface
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/EligibleStrategyContextFinder.java]: Verplaatsen naar methode

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        //foreach eligible context class
+        for (Pair<VariableDeclarator, ClassOrInterfaceDeclaration> eligibleContext : eligibleContexts) {
+            //Get the strategy interface of which the context type has a variable
+            ClassOrInterfaceDeclaration strategyInterface = eligibleContext.getValue();
+
+            //Get the variable deceleration
+            VariableDeclarator strategyVariable = eligibleContext.getKey();
+            ClassOrInterfaceType strategyInterfaceType = (ClassOrInterfaceType) strategyVariable.getType();
+
+            //Walk up the tree until we have the class containing the variable deceleration, this is the context class
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Typo

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            Node currentNode = strategyVariable;
+            while (!(currentNode instanceof ClassOrInterfaceDeclaration)) {
+                if (!currentNode.getParentNode().isPresent()) {
+                    break;
+                }
+                currentNode = currentNode.getParentNode().get();
+            }
+
+            //If we stopped but the current node is not a class we have a issue, so continue loop
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Ik snap deze zin niet helemaal

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+            //Find all method calls in the context class
+            List<MethodCallExpr> methodCalls = context.findAll(MethodCallExpr.class);
+
+            boolean classCallsStrategy = false;
+
+            //Loop over the method calls
+            for (MethodCallExpr methodCallExpr : methodCalls) {
+
+                //We are looking for a call in the scope of the variable deceleration
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: typo

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+                if (methodCallExpr.getScope().isPresent()) {
+
+                    //Resolve the type of the scope
+                    ResolvedType scopeType = javaSymbolSolver.calculateType(methodCallExpr.getScope().get());
+                    if (scopeType instanceof ReferenceTypeImpl) {
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Ik zie hier 3 geneste ifjes. Graag proberen te voorkomen.

Comment (Hawiak) [reply src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Dit kun je bijvoorbeeld doen door de if zo te schrijven dat hij eruit klapt als de situatie niet zo is als hij zou moeten zijn ipv checken of hij wel zo is.

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+                strategies.addAll(implementationFinder.getClasses());
+            }
+
+            //We should at least have one implementation of the strategy interface, else the pattern won't work
+            if (strategies.size() == 0) {
+                continue;
+            }
+
+            //At this point every requirement has been met so we make the Strategy class
+
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Zou deze lege regel weglaten

File: src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+            if (!(currentNode instanceof ClassOrInterfaceDeclaration)) {
+                continue;
+            }
+
+            ClassOrInterfaceDeclaration context = (ClassOrInterfaceDeclaration) currentNode;
+
+            //Find all method calls in the context class
+            List<MethodCallExpr> methodCalls = context.findAll(MethodCallExpr.class);
+
+            boolean classCallsStrategy = false;
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java]: Ik zal deze functionaliteit ook graag in een aparte methode zien.

File: src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            currentNode = currentNode.getParentNode().get();
+        }
+
+        //Set the ClassOrInterface part for the strategy context
+        if (currentNode instanceof CompilationUnit) {
+            if (((CompilationUnit) currentNode).getStorage().isPresent()) {
+                String filename = ((CompilationUnit) currentNode).getStorage().get().getFileName();
+                File file = new File(filename);
+
+                if (node.getRange().isPresent()) {
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/util/FilePartResolver.java]: 3 geneste ifjes, graag voorkomen

Zie https://stackoverflow.com/questions/268132/invert-if-statement-to-reduce-nesting

File: src/main/java/nl/windesheim/codeparser/analyzers/util/visitor/SetterFinder.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        for (Statement statement : body.getStatements()) {
+            if (statement instanceof ExpressionStmt) {
+                ExpressionStmt expressionStmt = (ExpressionStmt) statement;
+
+                Expression expression = expressionStmt.getExpression();
+                if (expression instanceof AssignExpr) {
+                    Expression target = ((AssignExpr) expression).getTarget();
+                    if (target instanceof FieldAccessExpr) {
+                        if (((FieldAccessExpr) target).getScope() instanceof ThisExpr) {
+                            if (((FieldAccessExpr) target).getName().equals(variable.getName())) {
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/util/visitor/SetterFinder.java]: 5 geneste ifjes, dit wordt heel lastig te testen en debuggen. Graag refactoren zodat je minder geneste ifjes krijgt.

File: src/main/java/nl/windesheim/codeparser/analyzers/util/visitor/SetterFinder.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        if (!optionalBody.isPresent()) {
+            return null;
+        }
+        BlockStmt body = optionalBody.get();
+
+        for (Statement statement : body.getStatements()) {
+            if (statement instanceof ExpressionStmt) {
+                ExpressionStmt expressionStmt = (ExpressionStmt) statement;
+
+                Expression expression = expressionStmt.getExpression();
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/util/visitor/SetterFinder.java]: Graag refactoreren, buiten de 5 geneste ifjes is het ook niet goed te testen via unit tests. Als je voor alle ifjes een aparte unit tests zou kunnen schrijven en er gaat wat kapot in de toekomst is het ook veel makkelijk terugvinden.

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * @return the deceleration
+     */
+    public ClassOrInterfaceDeclaration getDeceleration() {
+        return deceleration;
+    }
+
+    /**
+     * @param deceleration the deceleration to be set
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: typo

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+     * @return this
+     */
+    public ClassOrInterface setDeceleration(final ClassOrInterfaceDeclaration deceleration) {
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: typo

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * The filePart in which this class is found.
+     */
+    private FilePart filePart;
+
+    /**
+     * The AST node which was found.
+     */
+    private ClassOrInterfaceDeclaration deceleration;
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: typo

File: src/main/java/nl/windesheim/codeparser/ClassOrInterface.java [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    public ClassOrInterfaceDeclaration getDeceleration() {
+        return deceleration;
+    }
+
+    /**
+     * @param deceleration the deceleration to be set
+     * @return this
+     */
+    public ClassOrInterface setDeceleration(final ClassOrInterfaceDeclaration deceleration) {
+        this.deceleration = deceleration;
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/ClassOrInterface.java]: typo

File: src/main/java/nl/windesheim/codeparser/analyzers/singleton/SingletonAnalyzer.java [hunk 20]
```diff
@@ -79,16 +79,18 @@
 
                     Singleton singleton = new Singleton();
 
+
                     if (cu.getStorage().isPresent() && classDeclaration.getRange().isPresent()) {
```
Comment (Hawiak) [inline src/main/java/nl/windesheim/codeparser/analyzers/singleton/SingletonAnalyzer.java]: Graag de CU nog refactoren zodat het geen afkorting meer is.
